scripts/alert_emails/send_emails.ts

```diff
@@ -138,99 +46,104 @@ async function setLastSnapshotNumber(
-                );
-                await delay(rateLimitReset * 1000);
-              }
-              // Comment out this code if you are developing locally
-              await db
-                .collection(
-                  `snapshots/${currentSnapshot}/locations/${fips}/emails/`,
-                )
-                .doc(doc.id)
-                .set({
-                  sentAt: admin.firestore.FieldValue.serverTimestamp(),
-                });
-            })
-            .catch(async (err: CampaignMonitorError) => {
-              // if the email is invalid, move the document to the invalid collection
-              if (
-                err.Code === CM_INVALID_EMAIL_ERROR_CODE &&
-                err.Message === CM_INVALID_EMAIL_MESSAGE
-              ) {
-                invalidEmailCount += 1;
-                const currentData = (
-                  await db.collection('alerts-subscriptions').doc(doc.id).get()
-                ).data();
-                if (currentData) {
-                  await db
-                    .collection('invalid-alerts-subscriptions')
-                    .doc(doc.id)
-                    .set(currentData);
-                  await db
-                    .collection('alerts-subscriptions')
-                    .doc(doc.id)
-                    .delete();
-                }
-              } else {
-                errorCount += 1;
-              }
-            });
-        }
-      });
+      .get();
+    return querySnapshot.docs.map(emailDoc => emailDoc.id);
   }
+
+  async function onInvalidEmail(email: string) {
+    invalidEmailCount += 1;
+    const querySnapshot = await db
+      .collection('alerts-subscriptions')
+      .doc(email)
+      .get();
+    const currentData = querySnapshot.data();
+    if (currentData) {
+      await db
+        .collection('invalid-alerts-subscriptions')
+        .doc(email)
+        .set(currentData);
+      await db.collection('alerts-subscriptions').doc(email).delete();
+    }
+  }
+
+  async function onEmailSent(email: string, fips: string) {
+    emailSent += 1;
+    uniqueEmailAddress[email] = (uniqueEmailAddress[email] || 0) + 1;
+    locationsWithEmails[fips] = (locationsWithEmails[fips] || 0) + 1;
+    // comment out this code if you are developing locally
+    return db
+      .collection(`snapshots/${currentSnapshot}/locations/${fips}/emails/`)
+      .doc(email)
+      .set({ sentAt: admin.firestore.FieldValue.serverTimestamp() });
+  }
+
+  async function sendAlertEmail(email: string, fips: string) {
+    const locationAlert = alertsByLocation[fips];
+    const sendData = generateAlertEmailData(email, locationAlert);
+    if (dryRun) {
+      await onEmailSent(email, fips);
+      return;
+    }
+
+    try {
+      const response = await createSend.sendClassicEmail(sendData);
+      await onEmailSent(email, fips);
+      return response;
+    } catch (err) {
+      if (isInvalidEmailError(err)) {
+        await onInvalidEmail(email);
+        return;
```

Comment:
As an opinionated stylistic note, I vastly prefer if/else statements over early returns.  It's easy to overlook a "return" statement when reading the code and completely misunderstand the control-flow.  My one exception is for edge-case handling (e.g. error conditions) at the top of a function.  Then doing an early-return can make the rest of the function simpler to reason about.

But you're welcome to disregard my style preferences. 😁

Reply:
makes sense, updating!